codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/endpoints/EndpointResolverGenerator.java

```diff
@@ -0,0 +1,529 @@
+
+    private EndpointResolverGenerator(Builder builder) {
+        var parametersType = SmithyBuilder.requiredState("parametersType", builder.parametersType);
+        var resolverInterfaceType = SmithyBuilder.requiredState("resolverInterfaceType", builder.resolverInterfaceType);
+        var resolverImplementationType = SmithyBuilder.requiredState("resolverImplementationType",
+                builder.resolverImplementationType);
+        var newResolverFn = SmithyBuilder.requiredState("newResolverFn", builder.newResolverFn);
+        var endpointType = SmithyBuilder.requiredState("endpointType", builder.endpointType);
+        var resolveEndpointMethodName = SmithyBuilder.requiredState("resolveEndpointMethodName",
+                builder.resolveEndpointMethodName);
+
+        this.fnProvider = SmithyBuilder.requiredState("fnProvider", builder.fnProvider);
+        this.commonCodegenArgs = MapUtils.of(
+                "paramArgName", PARAMS_ARG_NAME,
+                "parametersType", parametersType,
+                "endpointType", endpointType,
+                "resolverInterfaceType", resolverInterfaceType,
+                "resolverImplementationType", resolverImplementationType,
+                "newResolverFn", newResolverFn,
+                "resolveEndpointMethodName", resolveEndpointMethodName,
+                "fmtErrorf", SymbolUtils.createValueSymbolBuilder("Errorf", SmithyGoDependency.FMT).build());
+    }
+
+    public GoWriter.Writable generate(Optional<EndpointRuleSet> ruleset) {
+        if (ruleset.isPresent()) {
+            return generateResolverType(generateResolveMethodBody(ruleset.get()));
+        } else {
+            LOGGER.warning("service does not have modeled endpoint rules");
+            return generateEmptyRules();
+        }
+    }
+
+    public GoWriter.Writable generateEmptyRules() {
+        return generateResolverType(generateEmptyResolveMethodBody());
+    }
+
+    private GoWriter.Writable generateResolverType(GoWriter.Writable resolveMethodBody) {
+        return goTemplate("""
+                // $resolverInterfaceType:T provides the interface for resolving service endpoints.
+                type $resolverInterfaceType:T interface {
+                    $resolveEndpointMethodDocs:W
+                    $resolveEndpointMethodName:L(ctx $context:T, options $parametersType:T) (
+                        $endpointType:T, error,
+                    )
+                }
+
+                $resolverTypeDocs:W
+                type $resolverImplementationType:T struct{}
+
+                func $newResolverFn:T() *$resolverImplementationType:T {
+                    return &$resolverImplementationType:T{}
+                }
+
+                $resolveEndpointMethodDocs:W
+                func (r *$resolverImplementationType:T) $resolveEndpointMethodName:L(
+                    ctx $context:T, $paramArgName:L $parametersType:T,
+                ) (
+                    endpoint $endpointType:T, err error,
```

Comment:
Do we really get anything out of using the predeclared return style here?

Reply:
none that i am aware of other than not including the declaration of `endpoint` in the body, if thats even a benefit. tbh, i didnt think too hard about this and let it be since i wasnt aware of any actionable downside or benefit